### PR TITLE
Add mojo service for accessing the system clipboard.

### PR DESCRIPTION
**Pull request by mpcomplete**
Hooked up for iOS and Android.

BUG=https://github.com/flutter/flutter/issues/1567

### CODE REVIEW

**Comment by mpcomplete** [comment]
@chinmaygarde might want to review the ios part

File: sky/build/sdk_xcode_harness/Local.xcconfig [hunk 1]
```diff
@@ -1,11 +0,0 @@
-// Copyright 2015 The Chromium Authors. All rights reserved.
```
Comment (chinmaygarde) [inline sky/build/sdk_xcode_harness/Local.xcconfig]: This file shouldn't be deleted.

File: sky/build/sdk_xcode_harness/Local.xcconfig [hunk 2]
```diff
@@ -1,11 +0,0 @@
...
-// Use of this source code is governed by a BSD-style license that can be
-// found in the LICENSE file.
-
-//
-// This file generated. Do not modify or check into version control.
-//
-
-DART_SDK_PATH=/path/to/dart/sdk
-FLUTTER_APPLICATION_PATH=path/to/flutter/application
-FLUTTER_ROOT=/path/to/flutter/root
```
Comment (abarth) [inline sky/build/sdk_xcode_harness/Local.xcconfig]: Oh noes!  Don't delete this file.  :)

File: sky/services/editing/editing.mojom [hunk 3]
```diff
@@ -48,3 +48,13 @@ interface Keyboard {
   Show();
   Hide();
 };
+
+struct ClipData {
```
Comment (abarth) [inline sky/services/editing/editing.mojom]: Maybe ClipboardData ?  ClipData makes me think we're going to visually limit the extent of something with a clip.

File: sky/services/editing/ios/clipboard_impl.mm [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ClipboardImpl::ClipboardImpl(
+    mojo::InterfaceRequest<::editing::Clipboard> request)
+    : binding_(this, request.Pass()) {}
+
+ClipboardImpl::~ClipboardImpl() {
+}
+
+void ClipboardImpl::SetClipData(::editing::ClipDataPtr clip) {
+  UIPasteboard* pasteboard = [UIPasteboard generalPasteboard];
+  pasteboard.string = [NSString stringWithUTF8String:clip->text.data()];
```
Comment (chinmaygarde) [inline sky/services/editing/ios/clipboard_impl.mm]: You dont have but you can replace the long form with `@(clip->text.data())`

File: sky/services/editing/editing.mojom [hunk 5]
```diff
@@ -48,3 +48,13 @@ interface Keyboard {
...
+  string text;
+};
+
+[ServiceName="editing::Clipboard"]
+interface Clipboard {
+  SetClipData(ClipData clip);
+  GetClipData() => (ClipData? clip);
```
Comment (abarth) [inline sky/services/editing/editing.mojom]: You probably want to add a `string format` argument to this function that lets the client say what sort of data it wants back from the clipboard (e.g., text, images, etc).  See https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/getData

Comment (abarth) [inline sky/services/editing/editing.mojom]: See also https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIPasteboard_Class/#//apple_ref/occ/instm/UIPasteboard/dataForPasteboardType:

File: sky/services/editing/src/org/domokit/editing/ClipboardImpl.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        clipboard.setPrimaryClip(clip);
+    }
+
+    @Override
+    public void getClipData(GetClipDataResponse callback) {
+        ClipboardManager clipboard =
+            (ClipboardManager) mContext.getSystemService(Context.CLIPBOARD_SERVICE);
+        android.content.ClipData clip = clipboard.getPrimaryClip();
+        if (clip == null ||
+            !clip.getDescription().hasMimeType(ClipDescription.MIMETYPE_TEXT_PLAIN)) {
```
Comment (abarth) [inline sky/services/editing/src/org/domokit/editing/ClipboardImpl.java]: Here again you see that the "get" cares about the type you're getting.

**Comment by abarth** [comment]
LGTM w/ mime types on get.  It's fine to only support text/plain for now, but we should structure the interface so that we can extend it in the future when we want to move images and such through the clipboard.

**Comment by eseidelGoogle** [comment]
Android used to only do text, but now can do more types:
http://developer.android.com/reference/android/content/ClipboardManager.html

File: sky/services/editing/editing.mojom (hunk 5 above)
Comment (mpcomplete) [inline sky/services/editing/editing.mojom]: My vision was to return whatever was in the clipboard, and the client could query what type it got back. I don't have a strong preference for doing it one way or the other. Is there a reason the `format` argument is superior?

Comment (abarth) [inline sky/services/editing/editing.mojom]: Yeah, there might be a large amount of data on the clipboard (e.g., an image or a video) that's not worth transferring if you really wanted some text.  Another use case is if there are multiple representations of the object on the clipboard (e.g., rich text or plain text), you only actually want to generate the representation that the client can understand.

Comment (mpcomplete) [inline sky/services/editing/editing.mojom]: OK, makes sense. How about an enum format rather than a string?

Comment (krisgiesing) [inline sky/services/editing/editing.mojom]: Using an enum means that the framework has to define all possible types. I can imagine some operations allowing for user-created type transfer where the sender and receiver agree on some uniquely identified format through the string argument, e.g. "com.example.album-schema-xml".  The framework could define string constants for certain well known types, e.g. text, image, URL, etc.

Comment (abarth) [inline sky/services/editing/editing.mojom]: You're example is right in the sense that vendors make up random mime types for their specific formats.

**Comment by abarth** [comment]
LGTM

**Pull request merged**